### ISSUE TITLE
Missing type error for invalid use of #private field

### ISSUE THREAD

**Issue by trusktr** [microsoft/TypeScript#43711]
# Bug Report

ES #private fields are [not source-positional](https://github.com/tc39/proposal-class-fields/issues/328), but scoped to each instance of a class. (That is unfortunate, in my humble opinion). However TypeScript is not detecting this case.

### 🔎 Search Terms

typescript no type error private field mixin

### 🕗 Version & Regression Information

<!-- When did you start seeing this bug occur?

4.3-beta (I didn't try earlier)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=99&ts=4.3.0-beta#code/C4TwDgpgBAwsD2AnAPAFSgXigbwL4D5MoA7CAdygAoA6WgQ0QHMBnALijuJAG0BdASkyFUAKBEAzAK7EAxsACW8YlABi8eAFl5AD3nFkMKBG3AIxACbNYCRPkoAhOswjsYg7CKhREEYJMTKMgA2TlZq8EYmZpZQjs44nl5QAMTi6kQAjABMAMyJXqbMwJRp8Ozh7vlJMkrM8EEQ1EHwjCXq1Knq-FW4ib29IsGhsHTAkaYWYepausSUAPIARgBWEHLuvUPMVgAiLePRU5o6egsra8AbYjXERVAyRKQUMKOD1IXFT1B7jPxAA)

### 💻 Code

```ts
type Ctor<T = {}> = new (...args: any[]) => T

function FooMixin<C extends Ctor>(Base: C) {
  return class Foo extends Base {
    #foo = 123
    test(foo: Foo) {
      console.log(foo.#foo)
    }
  }
}

class Cat extends FooMixin(Object) {}
class Dog extends FooMixin(Object) {}

const c = new Cat
c.test(new Dog) // Should there be a type error here, perhaps?
```

### 🙁 Actual behavior

No type error, but only a runtime error.

### 🙂 Expected behavior

There should perhaps also be a type error, if that's possible.

### Mixin docs

The [TypeScript Mixin documentation](https://www.typescriptlang.org/docs/handbook/mixins.html) advertises that people can use #private fields with class-factory mixins:

> ```ts
> // Mixins may not declare private/protected properties
> // however, you can use ES2020 private fields
> ```

This may lead people to the try them and they'll have this issue.

**Comment by DanielRosenwasser** [comment]
This is probably a design limitation. There's no way to distinguish whether any given utterance of a `Foo` is really a distinct one or an appropriate one. We have to get out of your way or else it would be impossible to call `new Dog().test(new Dog())`.

**Comment by trusktr** [comment]
I suppose it is possible, but complicating: I imagine there would need to be different `Foo` types associated with different function calls.

Honestly though, I'm not happy with TC39 introducing language incompatibilities all over the place.
